File: TD/03-tdp-dicho-knapsack/dichoIter.scala
def dicho(tab:Array[Int], elm:Int):Int = {
    var min = 0;
    var max = tab.length-1;
    while (max-min>1) {
        val milieu = (min+max) / 2;
        if (elm < tab(milieu)) {
            max = milieu;
        } else {
            min = milieu;
        }
    }
    if (tab(min) == elm) {
        return min
    } else if (tab(max) == elm) {
        return max
    } else {
        return -1
    }
}

def test(tab:Array[Int]) {
    def cherche(searched:Int, expected:Int) {
        val found = dicho(tab, searched);
        if (found != expected) 
            println("ERREUR: elm "+searched+" trouvé en "+found+" (attendu en "+expected+").");
        else
            println("elm "+searched+" trouvé en "+found+" comme attendu.");
    }
    cherche(0,-1);
    for (cpt <- 0 to tab.length-1) 
        cherche(tab(cpt),cpt);
    cherche(5,-1);
    cherche(11,-1);
}

test(Array(1,3,6,7,8,9,10))
test(Array(1,3,6,7,9,10))

File: TD/03-tdp-dicho-knapsack/src/dicho/DichoIter.java
public class DichoIter {
  public static int dichoIter(int tab[], int min, int max, int elem) {
    while (max - min > 1) {
      int milieu = (min + max) / 2;
      System.out.println("min: "+min+"; max: "+max+"; milieu: "+milieu);
      
      if (elem < tab[milieu]) {
        System.out.println("Debut ("+min+";"+milieu+")");
        max=milieu;
      } else {
        System.out.println("Fin   ("+milieu+";"+max+")");
        min=milieu;
      }
    }

    if (tab[min] == elem) {
      return min;
    } else if (tab[max] == elem) {
      return max;
    } else {
      return -1;
    }
  }
  public static void main(String[] args) {
     int tab[] = {1,3,6,7,8,9,10};
     int min=0,max=6,elem;
     System.out.println("RES: L'élément 0 se trouve en position "+
			dichoIter(tab,min,max,0)+" Attendu: -1");
     for (int cpt=0;cpt <=max; cpt++) {
       System.out.println("RES: L'élément "+tab[cpt]+" se trouve en position "+
			  dichoIter(tab,min,max,tab[cpt])+" Attendu: "+cpt);
     }
     System.out.println("RES: L'élément 5 se trouve en position "+ 
			dichoIter(tab,min,max,5)+" Attendu: -1");
     System.out.println("RES: L'élément 11 se trouve en position "+ 
			dichoIter(tab,min,max,11)+" Attendu: -1");
  }
}
